libs/saml2-web-sso/src/SAML2/WebSSO/API.hs

```diff
@@ -188,32 +227,33 @@ simpleVerifyAuthnResponse creds raw = do
 
 -- | Call verify and, if that fails, any work-arounds we have.  Discard all errors from
 -- work-arounds, and throw the error from the regular verification.
-allVerifies :: forall m err. (MonadError (Error err) m) => NonEmpty SignCreds -> LBS -> [String] -> m ()
+allVerifies :: forall m err. (MonadError (Error err) m) => NonEmpty SignCreds -> LBS -> NonEmpty String -> m (NonEmpty Assertion)
 allVerifies creds raw nodeids = do
-  let workArounds :: [Either String ()]
-      workArounds =
-        [ verifyADFS creds raw nodeids
-        ]
-  case verify creds raw `mapM_` nodeids of
-    Right () -> pure ()
-    Left err -> do
-      if any isRight workArounds
-        then pure ()
-        else throwError . BadSamlResponseInvalidSignature $ cs err
+  let workArounds = verifyADFS creds raw nodeids
+  xmls <- case verify creds raw `mapM` nodeids of
+    Right assertions -> pure assertions
+    Left err -> case workArounds of
+      Right ws -> pure ws
+      Left _ -> throwError . BadSamlResponseInvalidSignature $ cs err
+  (renderVerifyErrorHack . parseFromXmlTree) `mapM` xmls
+  where
+    -- (there must be a better way for this, but where?)
+    renderVerifyErrorHack :: forall m' err' a. (MonadError (Error err') m') => Either String a -> m' a
+    renderVerifyErrorHack = either throwError pure . fmapL (BadSamlResponseSamlError . LT.pack)
```

Comment:
Can this be resolved? What is the hack and is it acceptable?

Reply:
it's hacky because it goes through a string representation instead of deriving one symbolic representation from the other, but as long as serialization is intact, this is just ugly, not harmful or risky.